torchao/quantization/prototype/qat/api.py

```diff
@@ -18,20 +18,88 @@
     Int8DynActInt4WeightLinear,
     WeightOnlyInt4Linear,
 )
-from torchao.quantization.quant_primitives import ZeroPointDomain
+from torchao.quantization.linear_activation_quantized_tensor import (
+    to_linear_activation_quantized,
+)
+from torchao.quantization.quant_api import (
+    _get_linear_subclass_inserter,
+    _replace_with_custom_fn_if_matches_filter,
+    int4_weight_only,
+    int8_dynamic_activation_int4_weight,
+    quantize_,
+)
+from torchao.quantization.quant_primitives import (
+    MappingType,
+    ZeroPointDomain,
+)
 from torchao.quantization.unified import TwoStepQuantizer
-from torchao.quantization.utils import get_group_qparams_symmetric
+from torchao.quantization.utils import (
+    _get_per_token_block_size,
+    get_group_qparams_symmetric,
+)
+from .affine_fake_quantized_tensor import to_affine_fake_quantized
 from .utils import (
     _choose_qparams_per_token_asymmetric,
+    _enable_fake_quant,
     _fake_quantize_per_channel_group,
     _fake_quantize_per_token,
+    _is_linear_with_fq_weight,
+    _unwrap_affine_fake_quantized_tensor,
 )
 
 
 # =================
 # |   8da4w QAT   |
 # =================
 
+def int8_dynamic_activation_int4_weight_fake_quantize(group_size=32):
+    """
+    Applies int8 dynamic per token asymmetric activation fake quantization and
+    int4 per group weight symmetric fake quantization to linear. Please see
+    :func:`~torchao.quantization.int8_dynamic_activation_int4_weight` for more details.
+
+    Example usage:
+        from torchao.quantization import quantize_
+        quantize_(model, int8_dynamic_activation_int4_weight_fake_quantize(group_size=32))
+    """
+    def _apply_fake_quant(weight):
+        # avoid circular dep
+        from torchao.dtypes import to_affine_quantized
+
+        # weight settings
+        mapping_type = MappingType.SYMMETRIC
+        block_size = (1, group_size)
+        target_dtype = torch.int8
+        eps = torch.finfo(torch.float32).eps
+        quant_min = -8
+        quant_max = 7
+
+        # input settings
+        input_mapping_type = MappingType.ASYMMETRIC
+        input_target_dtype = torch.int8
+
+        def input_quant_func(x: torch.Tensor):
+            return to_affine_fake_quantized(
+                x,
+                input_mapping_type,
+                _get_per_token_block_size(x),
+                input_target_dtype,
+            )
+
+        weight = to_affine_fake_quantized(
```

Comment:
if your only goal is "my model has an `m.weight` is a tensor subclass, I want to be able to run the model, call `.backward()`, and get gradient computed directly on my subclass", then I think instead of this all you really need is:

```
weight = MySubclass(inner_tensor)
weight.requires_grad_(True)
model.weight = torch.nn.Parameter(weight)
```